machine/v1beta1/types_vsphereprovider.go

```diff
@@ -87,12 +87,124 @@ type NetworkSpec struct {
 	Devices []NetworkDeviceSpec `json:"devices"`
 }
 
+type DHCPState string
+
+const (
+	EnabledState  DHCPState = "Enabled"
+	DisabledState DHCPState = "Disabled"
+)
+
+// AddressesFromPool is an IPAddressPool that will be used to create
+// IPAddressClaims for fulfillment by an external controller.
+type AddressesFromPool struct {
+	Group string `json:"group"`
+
+	// Kind is the type of resource being referenced
+	Resource string `json:"resource"`
+
+	// Name is the name of resource being referenced
+	Name string `json:"name"`
+}
+
 // NetworkDeviceSpec defines the network configuration for a virtual machine's
 // network device.
 type NetworkDeviceSpec struct {
 	// NetworkName is the name of the vSphere network to which the device
 	// will be connected.
-	NetworkName string `json:"networkName"`
+	// +optional
+	NetworkName string `json:"networkName,omitempty"`
+
+	// DeviceName may be used to explicitly assign a name to the network device
+	// as it exists in the guest operating system.
+	// +optional
+	DeviceName string `json:"deviceName,omitempty"`
+
+	// DHCP4 is a flag that indicates whether or not to use DHCP for IPv4
+	// on this device.
+	// If enabled, then IPAddrs should not contain any IPv4 addresses and
+	// AddressesFromPools must be empty.
+	// +optional
+	// +kubebuilder:default=Enabled
+	// +kubebuilder:validation:Enum=Enabled;Disabled
+	DHCP4 DHCPState `json:"dhcp4,omitempty"`
+
+	// DHCP6 is a flag that indicates whether or not to use DHCP for IPv6
+	// on this device.
+	// If enabled, then IPAddrs should not contain any IPv6 addresses and
+	// AddressesFromPools must be empty.
+	// +optional
+	// +kubebuilder:default=Enabled
+	// +kubebuilder:validation:Enum=Enabled;Disabled
+	DHCP6 DHCPState `json:"dhcp6,omitempty"`
+
+	// Gateway4 is the IPv4 gateway used by this device.
+	// Required when DHCP4 is false.
+	// +optional
+	// +kubebuilder:validation:Format=ipv4
+	Gateway4 string `json:"gateway4,omitempty"`
+
+	// Gateway4 is the IPv4 gateway used by this device.
+	// Required when DHCP6 is false.
+	// +kubebuilder:validation:Format=ipv6
+	// +optional
+	Gateway6 string `json:"gateway6,omitempty"`
+
+	// IPAddrs is a list of one or more IPv4 and/or IPv6 addresses and CIDR to assign
+	// to this device.  For example: 192.168.1.100/24
+	// Required when DHCP4 and DHCP6 are both Disabled.
+	// +optional
+	IPAddrs []string `json:"ipAddrs,omitempty"`
+
+	// MTU is the device’s Maximum Transmission Unit size in bytes.
```

Comment:
Need to make sure to explain maximum, minimum and default values in prose in the godoc please